ext/websocket/lib.rs

```diff
@@ -500,24 +499,40 @@ pub async fn op_ws_next_event(
   let cancel = RcRef::map(&resource, |r| &r.cancel);
   let val = resource.next_message(cancel).await?;
   let res = match val {
-    Some(Ok(Message::Text(text))) => NextEventResponse::String(text),
-    Some(Ok(Message::Binary(data))) => NextEventResponse::Binary(data.into()),
-    Some(Ok(Message::Close(Some(frame)))) => NextEventResponse::Close {
-      code: frame.code.into(),
-      reason: frame.reason.to_string(),
-    },
-    Some(Ok(Message::Close(None))) => NextEventResponse::Close {
-      code: 1005,
-      reason: String::new(),
-    },
-    Some(Ok(Message::Ping(_))) => NextEventResponse::Ping,
-    Some(Ok(Message::Pong(_))) => NextEventResponse::Pong,
-    Some(Err(e)) => NextEventResponse::Error(e.to_string()),
+    Some(Ok(Message::Text(text))) => {
+      (MessageKind::Text as u16, StringOrBuffer::String(text))
+    }
+    Some(Ok(Message::Binary(data))) => (
+      MessageKind::Binary as u16,
+      StringOrBuffer::Buffer(data.into()),
+    ),
+    Some(Ok(Message::Close(Some(frame)))) => (
+      frame.code.into(),
+      StringOrBuffer::String(frame.reason.to_string()),
+    ),
+    Some(Ok(Message::Close(None))) => {
+      (1005, StringOrBuffer::String("".to_string()))
+    }
+    Some(Ok(Message::Ping(_))) => (
+      MessageKind::Ping as u16,
+      StringOrBuffer::Buffer(vec![].into()),
```

Comment:
Does it make sense to send these empty buffers? Could we replace them with `undefined`? That still wouldn't address the fact that this function is megamorphic and will get deoptimized by V8 every time data received is different than when the function got jitted (if it gets jitted at all)

Reply:
Idea to explore: store the received data on the Websocket struct, return only the number to JS and immediately call a sync op that either gets a buffer or a string (ie. two separate ops). That way we wouldn't make V8 deoptimize this function

Reply:
> Does it make sense to send these empty buffers? Could we replace them with undefined? 

`Option<StringOrBuffer>` makes it a little slow, since Ping and Pong are frequent events i've left them as-is.

> That still wouldn't address the fact that this function is megamorphic and will get deoptimized by V8 every time data received is different than when the function got jitted (if it gets jitted at all)

It won't get deoptimized everytime return value changes. AFAIK V8 can learn that this function returns either a AB or string back. It'll have JIT code prepared to handle that after a few calls.